src/gfx/util.zig

```diff
@@ -0,0 +1,191 @@
+            result.next_collision_index = sparse_vertices_count;
+            result.next_packed_index = 0;
+            result.written_indices_count = 0;
+            std.mem.set(MapEntry, result.sparse_to_packed_map, .{});
+            return result;
+        }
+
+        pub fn put(self: *@This(), vertex: Vertex, sparse_index: IndexType) void {
+            if (self.sparse_to_packed_map[sparse_index].packed_index == null_index) {
+                // New start of chain, reserve a new packed index and add entry to `index_map`
+                const packed_index = self.next_packed_index;
+                self.sparse_to_packed_map[sparse_index].packed_index = packed_index;
+                self.vertices[packed_index] = vertex;
+                self.indices[self.written_indices_count] = packed_index;
+                self.written_indices_count += 1;
+                self.next_packed_index += 1;
+                return;
+            }
+            var previous_sparse_index: IndexType = undefined;
+            var current_sparse_index = sparse_index;
+            while (current_sparse_index != null_index) {
+                const packed_index = self.sparse_to_packed_map[current_sparse_index].packed_index;
+                if (std.mem.eql(u8, &std.mem.toBytes(self.vertices[packed_index]), &std.mem.toBytes(vertex))) {
+                    // We already have a record for this vertex in our chain
+                    self.indices[self.written_indices_count] = packed_index;
+                    self.written_indices_count += 1;
+                    return;
+                }
+                previous_sparse_index = current_sparse_index;
+                current_sparse_index = self.sparse_to_packed_map[current_sparse_index].next_sparse;
+            }
+            // This is a new mapping for the given sparse index
+            const packed_index = self.next_packed_index;
+            const remapped_sparse_index = self.next_collision_index;
+            self.indices[self.written_indices_count] = packed_index;
+            self.vertices[packed_index] = vertex;
+            self.sparse_to_packed_map[previous_sparse_index].next_sparse = remapped_sparse_index;
+            self.sparse_to_packed_map[remapped_sparse_index].packed_index = packed_index;
+            self.next_packed_index += 1;
+            self.next_collision_index += 1;
+            self.written_indices_count += 1;
+        }
+
+        pub fn deinit(self: *@This(), allocator: std.mem.Allocator) void {
+            allocator.free(self.vertices);
+            allocator.free(self.indices);
+            allocator.free(self.sparse_to_packed_map);
+        }
+
+        pub fn indexBuffer(self: @This()) []IndexType {
+            return self.indices;
+        }
+
+        pub fn vertexBuffer(self: @This()) []Vertex {
+            return self.vertices[0..self.next_packed_index];
+        }
+    };
+}
+
+const Vec4 = [4]f32;
+const Vec3 = [3]f32;
+const Vec2 = [2]f32;
```

Comment:
wouldn't be better to use `@Vector`?

Reply:
Could be, although this is only a test. I guess it would come down to which is easier to read / a more common use-case. I don't have any issues changing to @Vector.

Reply:
An issue to keep in mind with `@Vector` is that `@Vector(3, T)` has the same alignment/padding requirements as `@Vector(4, T)`, so it is not necessarily ideal for vertex buffers.